### ISSUE TITLE
Utilities for diagnosing diffusive CFL numbers.

### ISSUE THREAD

**Issue by ali-ramadhan** [climate-machine/Oceananigans.jl#360]
Proposing to add this utility back in as it seems very useful when running with LES closures such as AMD.

It will have to dispatch on the `TurbulenceClosure` type, but there aren't that many types.

**Comment by glwagner** [comment]
I'm just using

```julia
function cell_diffusion_timescale(model)
    Δ = min(model.grid.Δx, model.grid.Δy, model.grid.Δz)
    max_ν = maximum(model.diffusivities.νₑ.data.parent)
    max_κ = max(Tuple(maximum(κₑ.data.parent) for κₑ in model.diffusivities.κₑ)...)
    return min(Δ^2 / max_ν, Δ^2 / max_κ)
end
```

On #new-closures, I've moved the type parameter associated with the closure to 2nd in line (after the time-stepper), which makes it easier to dispatch on.

Might be a good idea to define a `maximum` function for fields so we don't have to do the `.data.parent` shenanigans.

Obviously its easy to add this to `utils.jl`. The main con, which applies to lots of these little features added to our unstable, rapidly changing code, is that we have to maintain it.

**Comment by glwagner** [comment]
Are we sure this is very useful, or is it simply a hypothesis that this may be useful?

**Comment by ali-ramadhan** [comment]
Thanks for the code snippet!

Hmmm, maybe it's worth holding off for a bit, but I think if we find it to be a useful utility in verifying the model and LES closures, then it should be added in and maintained.

Certainly it's useful in diffusion-dominated simulations, however they are the exception rather than the norm (although I have been running some diffusion-dominated 1D models).

**Comment by glwagner** [comment]
@ali-ramadhan, here's one idea:

```julia
mutable struct CFLDiagnostic{DT, TS}
    cfl :: Float64
    Δt :: DT   
    frequency :: Int 
    timescale :: TS
end

AdvectiveCFLDiagnostic(Δt; frequency=1) = 
    CFLDiagnostic(Δt, frequency, cell_advection_timescale)

DiffusiveCFLDiagnostic(Δt; frequency=1) = 
    CFLDiagnostic(Δt, frequency, cell_diffusion_timescale)

# Default
CFLDiagnostic(Δt; frequency=1) = AdvectiveCFLDiagnostic(Δt; frequency=frequency)

(diag::CFLDiagnostic{<:Number})(model) = diag.Δt / c.timescale(model)
(diag::CFLDiagnostic{<:TimeStepWizard})(model) = diag.Δt.Δt / c.timescale(model)

function run_diagnostic(model, diag::CFLDiagnostic) 
    diag.cfl = cfl(model)
    return
end
```

We can then write `cell_diffusion_timescale` to dispatch on the type of `model.closure`.

**Comment by ali-ramadhan** [comment]
Nice!

I was actually going to advocate for adding such a utility because I've found myself running a few tests where the time steps are likely to be diffusion limited:
1. possibly stratified Couette flow sometimes,
2. the Pearson vortex test from PR #336,
3. more recently the [forced flow test](https://github.com/climate-machine/Oceananigans.jl/blob/5355044deae3120134b55cc20f6e38f637eea6a5/test/verification/forced_flow.jl) from (Brown, Cortez, & Minion, 2000) that I'm using to test whether our operator-splitting method is second-order accurate in both velocity and pressure.

For geophysical flows, the time step will almost always be limited by advection, but there seem to be many other flows where the time step will be limited by diffusion. So if we want a package that is usable for not just geophysical flows, it would be nice to have adaptive time stepping that works with both advective and diffusive CFL.

**Comment by glwagner** [comment]
@ali-ramadhan, great!

For the record:

I think that diffusion can also limit time steps in 'geophysical' flows (despite that this doesn't seem to be the case in the very small number of examples we have so far considered in Oceananigans). Many models use high-order diffusivity, for example, which scales with higher-than-quadratic powers of the grid spacing and can become an issue at high resolution. In idealized geophysical applications, and especially with spectral methods 8th, and even 12th power diffusion operators are often used. For these applications, an implicit treatment of diffusion terms is essential for numerical stability. 

In "realistic" oceanographic applications, grids are often extremely anisotropic with very fine vertical spacing, in which case vertical diffusion can become rate-limiting. For this reasons, many ocean codes treat vertical diffusion implicitly (which incurs a relatively cheap tridiagonal solve every time-step), while treating the less-stiff horizontal diffusion terms explicitly.

More generally, because the diffusive CFL increases quadratically with grid spacing, diffusion becomes more of a factor the more "resolved" the flow is. In LES the flow is only partially resolved; in addition the diffusive CFL formally scales identically as the advective CFL, so that I think we can typically expect them to have similar orders of magnitude. Thus for LES it seems that implicit treatment of diffusion terms / monitoring of the diffusive CFL is not crucial (though it will still be useful to have these tools available I think). Some oceanographic applications are a sort of "pseudo-LES" and also fall into this category.

**Comment by glwagner** [comment]
This issue was resolved by #407.
